include/deal.II/grid/connectivity.h

```diff
@@ -0,0 +1,1477 @@
+
+              for (unsigned int i = 0; i < local_entity_vertices.size(); i++)
+                entity_vertices[i] =
+                  cell_vertice[local_entity_vertices[i]] + offset;
+
+              // ... create key
+              std::array<unsigned int, key_length> key = entity_vertices;
+              std::sort(key.begin(), key.end());
+              keys.emplace_back(key, counter++);
+
+              additional_info.emplace_back(entity_vertices,
+                                           cell_type->type_of_entity(d, e));
+
+              if (comptibility_mode)
+                additional_info_compatibility.emplace_back(
+                  second_key_function(entity_vertices, cell_type, c, e));
+            }
+        }
+
+      col_d.resize(keys.size());
+      orientations.resize(keys.size());
+
+      // step 2: sort according to key so that entities with same key can be
+      // merged
+      std::sort(keys.begin(), keys.end());
+
+
+      if (comptibility_mode)
+        {
+          std::array<unsigned int, key_length> ref_key, new_key;
+          std::fill(ref_key.begin(), ref_key.end(), 0);
+          for (unsigned int i = 0; i < keys.size(); ++i)
+            {
+              const auto offset_i = std::get<1>(keys[i]);
+
+              if (ref_key != std::get<0>(keys[i]))
+                {
+                  ref_key = std::get<0>(keys[i]);
+                  new_key = additional_info_compatibility[offset_i];
+                }
+
+              std::get<0>(keys[i]) = new_key;
+            }
+
+          std::sort(keys.begin(), keys.end());
+        }
+
+      std::array<unsigned int, key_length> ref_key;
+      std::array<unsigned int, key_length> ref_indices;
+      std::fill(ref_key.begin(), ref_key.end(), 0);
+
+      for (unsigned int i = 0, counter = dealii::numbers::invalid_unsigned_int;
+           i < keys.size();
+           i++)
+        {
+          const auto offset_i = std::get<1>(keys[i]);
+
+          if (ref_key != std::get<0>(keys[i]))
+            {
+              // new key
+              counter++;
+              ref_key     = std::get<0>(keys[i]);
+              ref_indices = std::get<0>(additional_info[offset_i]);
+
+              ptr_0.push_back(col_0.size());
+              for (const auto j : std::get<0>(additional_info[offset_i]))
+                if (j != 0)
+                  col_0.push_back(j - offset);
+
+              // take its orientation as default
+              col_d[offset_i]        = counter;
+              orientations[offset_i] = 1;
+            }
+          else
+            {
+              col_d[offset_i] = counter;
+              orientations[offset_i] =
+                compute_orientation(ref_indices,
+                                    std::get<0>(additional_info[offset_i]),
+                                    std::get<1>(additional_info[offset_i]));
+            }
+        }
+      ptr_0.push_back(col_0.size());
```

Comment:
I like this function because it is very compact. You told me that the algorithm (here and/or for build_intersection below) is inspired by FEniCS's DOLFIN implementation. If that inspired here (or elsewhere in the file) you could probably mention it because it would allow others to read up about the background.